### PR TITLE
Add a Framebuffer cache

### PR DESCRIPTION
**Pull request by reduz**
Adds a FramebufferCache singleton that operates the same way as UniformSetCache. This is required for the rendering reorganization work.

Allows creating framebuffers on the fly (and keep them cached if re-requested) such as:

```C++
RID fb = FramebufferCache::get_singleton()->get_cache(texture1,texture2);
```

As plus, fixed a potential crash bug in UniformSetCache and cost corrected RID_Owner.

### CODE REVIEW

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 1]
```diff
@@ -4181,6 +4181,23 @@ RenderingDevice::FramebufferFormatID RenderingDeviceVulkan::framebuffer_get_form
 	return framebuffer->format_id;
 }
 
+bool RenderingDeviceVulkan::framebuffer_is_valid(RID p_framebuffer) const {
+	_THREAD_SAFE_METHOD_
+
+	const Framebuffer *framebuffer = framebuffer_owner.get_or_null(p_framebuffer);
```
Comment (RandomShaper) [inline drivers/vulkan/rendering_device_vulkan.cpp:4187]: If this function is the only reason to make the RID owner `mutable`, I'd rather do one of the following:
a) Go filthy locally, with a `const_cast` here.
b) Even better, use `RID_Owner::owns()`, which would require to make it `const`.

Comment (reduz) [reply drivers/vulkan/rendering_device_vulkan.cpp:4187]: You are right, I forgot we have owns(), will change it. Thanks!

Comment (reduz) [reply drivers/vulkan/rendering_device_vulkan.cpp]: It seems owns was not const, so I had to fix this.

File: servers/rendering/renderer_rd/framebuffer_cache_rd.h [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+		Vector<RID> textures;
+		_create_args(textures, args...);
+
+		return _allocate_from_data(1, h, table_idx, textures, Vector<RD::FramebufferPass>());
+	}
+
+	template <typename... Args>
+	RID get_cache_multiview(uint32_t p_views, Args... args) {
+		uint32_t h = hash_murmur3_one_32(p_views);
```
Comment (RandomShaper) [inline servers/rendering/renderer_rd/framebuffer_cache_rd.h:229]: Not important, but an `ERR_FAIL_COND_V(p_views >= 2, RID())` may help catch some errors?

File: servers/rendering/renderer_rd/framebuffer_cache_rd.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+	if (p_cache->next) {
+		p_cache->next->prev = p_cache->prev;
+	}
+
+	cache_allocator.free(p_cache);
+	cache_instances_used--;
+}
+void FramebufferCacheRD::_framebuffer_invalidation_callback(void *p_userdata) {
+	singleton->_invalidate(reinterpret_cast<Cache *>(p_userdata));
```
Comment (RandomShaper) [inline servers/rendering/renderer_rd/framebuffer_cache_rd.cpp:52]: `static_cast` is enough from `void *`.

**Review by RandomShaper — APPROVED** [review]
Looks good. I haven't spotted any issues in the complex parts of the code.

File: servers/rendering/renderer_rd/framebuffer_cache_rd.h (hunk 2 above)
Comment (reduz) [reply servers/rendering/renderer_rd/framebuffer_cache_rd.h:229]: I think you can technically have as many views as you want, if you restrict to 2 then its more like an implementation dependant thing.

File: servers/rendering/renderer_rd/framebuffer_cache_rd.cpp (hunk 3 above)
Comment (reduz) [reply servers/rendering/renderer_rd/framebuffer_cache_rd.cpp:52]: I have not much consensus on these ones, but I generally prefer to use static_cast for actual classes.

**Comment by reduz** [comment]
@RandomShaper thanks for the review!

**Review by BastiaanOlij — APPROVED** [review]
This looks really cool, can't wait to use it

**Comment by RandomShaper** [comment]
Some day we need to harmonize `Vector` and `LocalVector` in terms of the integer index type...

**Comment by reduz** [comment]
@RandomShaper indeed, we should move the index from Vector, from int to uint64_t.

**Comment by clayjohn** [comment]
Looks good to me too. Merging now so @BastiaanOlij has it available for his PR

**Pull request merged**